### PR TITLE
Distillation support for torchvision script

### PR DESCRIPTION
**Pull request by rahul-tuli**
The goal of this PR is to add distillation support to our `pytorch/torchvision` integration
Test recipe:

`distillation.yaml`:

```yaml
num_epochs: &num_epochs 10
lr: &lr 0.008

training_modifiers:
  - !EpochRangeModifier
    start_epoch: 0.0
    end_epoch: *num_epochs

  - !SetLearningRateModifier
    start_epoch: 0.0
    learning_rate: *lr

  - !DistillationModifier
     start_epoch: 0.0
     hardness: 0.5
     temperature: 2.0
     distill_output_keys: [0]

```

Test commands, (run manually):

- [X] self distillation

```bash
sparseml.image_classification.train \ 
    --recipe distillation.yaml --pretrained True --pretrained-dataset imagenette \
    --arch-key resnet50 --dataset-path /home/rahul/datasets/imagenette/imagenette-160 \
    --batch-size 128 --opt SGD --output-dir ./training-runs/image_classification-pretrained \
    --distill-teacher self
```

- [X] This command should fail (distillation recipe given but no `--distill-teacher` specified)

```bash
sparseml.image_classification.train \ 
    --recipe distillation.yaml --pretrained True --pretrained-dataset imagenette \
    --arch-key resnet50 --dataset-path /home/rahul/datasets/imagenette/imagenette-160 \
    --batch-size 128 --opt SGD --output-dir ./training-runs/image_classification-pretrained 
```

- [X] Disable distillation

```bash
sparseml.image_classification.train \ 
    --recipe distillation.yaml --pretrained True --pretrained-dataset imagenette \
    --arch-key resnet50 --dataset-path /home/rahul/datasets/imagenette/imagenette-160 \
    --batch-size 128 --opt SGD --output-dir ./training-runs/image_classification-pretrained \
    --distill-teacher disable
```

- [X] Distill a `mobilenet` using a `resnet50` teacher from sparsezoo

```bash
sparseml.image_classification.train \ 
    --recipe distillation.yaml --pretrained True --pretrained-dataset imagenette \
    --arch-key mobilenet --dataset-path /home/rahul/datasets/imagenette/imagenette-160 \
    --batch-size 128 --opt SGD --output-dir ./training-runs/image_classification-pretrained \
    --distill-teacher zoo:cv/classification/resnet_v1-50/pytorch/sparseml/imagenet/base-none \
    --pretrained-teacher-dataset imagenet --teacher-arch-key resnet50
```

### CODE REVIEW

**Review by corey-nm — CHANGES_REQUESTED** [review]
changes look good, but need to call manager.update_loss (or whatever the call is) to actually use distillation loss

**Comment by rahul-tuli** [comment]
> changes look good, but need to call manager.update_loss (or whatever the call is) to actually use distillation loss

Great catch, updated!

File: src/sparseml/pytorch/torchvision/train.py [hunk 1]
```diff
@@ -81,13 +83,24 @@ def train_one_epoch(
         start_time = time.time()
         image, target = image.to(device), target.to(device)
         with torch.cuda.amp.autocast(enabled=scaler is not None):
-            output = model(image)
+            outputs = output = model(image)
```
Comment (KSGulin) [inline src/sparseml/pytorch/torchvision/train.py:86]: Is this a typo or is the output getting assigned to two vars on purpose?

**Review by KSGulin — DISMISSED** [review]
Left one comment, otherwise LGTM!

**Review by bfineran — DISMISSED** [review]
Looks great @rahul-tuli

**Review by corey-nm — APPROVED** [review]

File: src/sparseml/pytorch/torchvision/train.py [hunk 2]
```diff
@@ -81,13 +83,24 @@ def train_one_epoch(
...
             if isinstance(output, tuple):
                 # NOTE: sparseml models return two things (logits & probs)
                 output = output[0]
             loss = criterion(output, target)
 
         if steps_accumulated % args.gradient_accum_steps == 0:
+            if manager is not None:
+                manager.loss_update(
```
Comment (corey-nm) [inline src/sparseml/pytorch/torchvision/train.py:94]: @rahul-tuli does loss_update return the new loss?

**Review by corey-nm — CHANGES_REQUESTED** [review]
Looks like loss_update returns the new loss to use! So close 😀

File: src/sparseml/pytorch/torchvision/train.py (hunk 1 above)
Comment (corey-nm) [reply src/sparseml/pytorch/torchvision/train.py:86]: Damn that is clever, nice! 🚀

**Review by corey-nm — APPROVED** [review]
🚀 LETS GOOOOOO

**Review by bfineran — DISMISSED** [review]

**Review by KSGulin — DISMISSED** [review]

File: src/sparseml/pytorch/torchvision/train.py [hunk 3]
```diff
@@ -91,13 +92,24 @@ def train_one_epoch(
...
             loss = criterion(output, target)
 
         if steps_accumulated % accum_steps == 0:
+            if manager is not None:
+                loss = manager.loss_update(
+                    loss=loss,
+                    module=model,
+                    optimizer=optimizer,
+                    epoch=epoch,
+                    steps_per_epoch=len(data_loader),
```
Comment (corey-nm) [inline src/sparseml/pytorch/torchvision/train.py:108]:
```suggestion
                    steps_per_epoch=len(data_loader) / accum_steps,
```

File: src/sparseml/pytorch/torchvision/train.py [hunk 4]
```diff
@@ -91,13 +92,24 @@ def train_one_epoch(
...
-            output = model(image)
+            outputs = output = model(image)
             if isinstance(output, tuple):
                 # NOTE: sparseml models return two things (logits & probs)
                 output = output[0]
             loss = criterion(output, target)
 
         if steps_accumulated % accum_steps == 0:
+            if manager is not None:
+                loss = manager.loss_update(
```
Comment (corey-nm) [inline src/sparseml/pytorch/torchvision/train.py:103]: Question: It looks like this is only called once per gradient accum step, instead of once per batch. Is this correct?

Comment (rahul-tuli) [reply src/sparseml/pytorch/torchvision/train.py:103]: Yes it needs to be called once per backward pass

Comment (corey-nm) [reply src/sparseml/pytorch/torchvision/train.py:103]: 👍

**Review by corey-nm — APPROVED** [review]

**Review by KSGulin — APPROVED** [review]
